### ISSUE TITLE
Deep Classification, Embedding & Text Generation - Devlin, Chang, Lee & Toutanova 2018

### ISSUE THREAD

**Issue by jamesallenevans** [Computational-Content-Analysis-2020/Readings-Responses#39]
Devlin, Jacob, Ming-Wei Chang, Kenton Lee, and Kristina Toutanova. 2018. [“Bert: Pre-training of deep bidirectional transformers for language understanding."](https://arxiv.org/pdf/1810.04805.pdf)” arXiv:1810.04805.

**Comment by lkcao** [comment]
I have tried to understand this very technical piece. I am a little curious about what makes an algorithm 'innovative' in computer science. Can we say BERT is basically a process that pre-train the parameters on a large bi-directional structure, and transfer these parameters in later tasks (either by feature-based approach or fine tuning) ? From this paper, it seems that before this paper many people have pre-trained parameters and do transfer learning on unidirectional structure, so the innovation of BERT is mainly from changing the unidirectional structure to bidirectional, is that right?

**Comment by katykoenig** [comment]
@clk16 I, too, had difficulty parsing this paper as it is high technical but understand that it is foundational due to its bidirectionality. After some more quick research on BERT, it seems that bidirectionality wasn't a new a idea but was previously impossible to implement at scale due to the compute power necessary until cloud TPUs were invented (see: https://ai.googleblog.com/2018/11/open-sourcing-bert-state-of-art-pre.html). This post also notes that BERT is available for public use using Google Colab. I am wondering if our weekly assignments could also benefit from the free TPU/GPU available on Colab (as well as our unlimited storage in our Google drives through our student accounts).

I also was a bit confused by the evaluation of BERT and other language representation models. Specifically, the paper states that BERT achieves higher F1 scores than a human, but how that is possible when the true labels are given by humans themselves (shouldn't human F1 be 100%)?

**Comment by bjcliang-uchi** [comment]
@katykoenig My take is that the true labels are certain aggregated human answers (like what we learned at the beginning of this class: human graders do not often agree with each other). 

My question is more about the application of BERT. It seems that BERT is designed to use in google search and translation. But is it possible to use BERT to summarize articles based on the embeddings?

**Comment by arun-131293** [comment]
> @clk16 I, too, had difficulty parsing this paper as it is high technical but understand that it is foundational due to its bidirectionality. After some more quick research on BERT, it seems that bidirectionality wasn't a new a idea but was previously impossible to implement at scale due to the compute power necessary until cloud TPUs were invented (see: https://ai.googleblog.com/2018/11/open-sourcing-bert-state-of-art-pre.html). This post also notes that BERT is available for public use using Google Colab. I am wondering if our weekly assignments could also benefit from the free TPU/GPU available on Colab (as well as our unlimited storage in our Google drives through our student accounts).
> 
> I also was a bit confused by the evaluation of BERT and other language representation models. Specifically, the paper states that BERT achieves higher F1 scores than a human, but how that is possible when the true labels are given by humans themselves (shouldn't human F1 be 100%)?

Compute power is not the only main issue with bidirectionally in language models. The conceptual issue is when you go predicting words left to right, you can’t also go right to left on the same text, since the text(each word and it’s context) is already seen. So, one would be just repeating the same task with no new information added. This is what the paper means when it states: “Unfortunately, standard conditional language models can only be trained left-to-right or right-to-left, since bidirectional conditioning would allow each word to indirectly “see itself”, and the model could trivially predict the target word in a multi-layered context”. 
Their masked LM approach overcomes this issue by dropping 15 pct random words in every sequence and only predicting them. This is the “masked LM” approach- one of the main innovations here. 
Do you mean to say the masked LM approach was already known before?

**Comment by laurenjli** [comment]
The reading states that "Fine-tuning is straightforward since the self attention mechanism in the Transformer allows BERT to model many downstream tasks." What is the self attention mechanism? It is mentioned a couple of times but I’m not clear on what it does. The section on self attention in the fundamental reading (Ch 10) is blank.

**Comment by heathercchen** [comment]
This paper is quite technical and difficult to understand. I am wondering why the authors proposed a bidirectional encoder for texts here. Since as we all know, the texts, at least English texts, are arranged in a way to read write from left to right, thus unidirectional. I am wondering about the linguistic intuition behind the proposal of this method, rather than just for improving accuracy.

**Comment by arun-131293** [comment]
> This paper is quite technical and difficult to understand. I am wondering why the authors proposed a bidirectional encoder for texts here. Since as we all know, the texts, at least English texts, are arranged in a way to read write from left to right, thus unidirectional. I am wondering about the linguistic intuition behind the proposal of this method, rather than just for improving accuracy.

It is true that, there is a difference in [linguistics between externalized and internal language](https://www.sciencedirect.com/science/article/pii/S1364661315002326). What you are referring to is the externalized language which has a temporal dimension. But internal language (so say, sentences in the mind) are conceptualized to have a tree structure which is neither left to right or right to left. 

Having said that, the idea of using bidirectionality in language models has little to do with linguistics and in general NLP as a field does not engage with classical lingusitcs(as the paper linked above written by Everaert, Chomsky and others notes). Rather, its simply based on the intution that you will be able to predict a word in a sentence better through statistical matching if you know both the left and right context.

Consider this example:
"As ____ walked up the stairs, John thought to himself it was easier climbing down."
Even though English is externalized left to right, to fill in the blank with "he", it will be useful to have the right context which contains the resolution of the pronoun("he" to "John").

**Comment by jsmono** [comment]
The paper presented a new way to approach the language representation model using Wikipedia and book. My question is not directly related to the paper, but I'm just curious if using other datasets, will the result of the training be different. Materials on Wikipedia and book are in general more formal and professional. So if we use more casual texts such as tweets and chat records, will this influence the result of the training? Will the generated words be more "casual" than "formal"?

**Comment by HaoxuanXu** [comment]
I'd love to know what self-attention means in the context of BERT and GPT. Does it refer to the implementation of the forget gate?

**Comment by wunicoleshuhui** [comment]
I'm also quite confused about the significance of the masked LM procedure. Why does masking some of the input tokens help with training the bidirectional model better? For what reasons it set at the 15% level, or was this more done arbitrarily?

**Comment by luxin-tian** [comment]
I hope I would understand technical papers like this in the future. As it seems the technology renders the possibility of artificial intelligence that assists humans with many tasks, I wonder what research in the field of social science would look like using such techniques?

**Comment by chun-hu** [comment]
I'd also like to learn more about how the self-attention mechanism applies here. We read about attention mechanisms in our fundamental reading -- are they referring to the same thing?

**Comment by sunying2018** [comment]
This paper is very innovative to break unidirectionality constraint by using a "masked language model".  I am quite confused about this procedure. As mentioned in the paper, they masked 15% of all WordPiece tokens in each sequence at random in all of experiments.  I am wondering why they need to keep this percent constant at 15% level and why they choose such percentage?

**Comment by alakira** [comment]
As many have pointed out, this paper does not describe the technical details of the attention mechanism used with encoders. Also, after choosing 15% to mask tokens, why they did some tricks. They explained that it was necessary to resolve the mismatch between the pre-training phase and the fine-tuning phase, but what is the mismatch?

**Comment by skanthan95** [comment]
> Critically, however, the BERT Transformer uses
bidirectional self-attention, while the GPT Transformer uses constrained self-attention where every
token can only attend to context to its left.

I tried to search for a clear definition of self-attention in this article and couldn't find one. What does this term mean in this context? How does it compare to a transformer that doesn't use self-attention, in what contexts would you want that to each sort of transformer?

**Comment by ziwnchen** [comment]
From my understanding, BERT is a self-supervised language model based on randomly blank tokens. As others mentioned, the advantage of BERT may come from its bidirectionality. But is it possible to have a more interpretable explanation of the performance improvement caused by the specific structures?
